### DISCUSSION TITLE
[v5] entry and invoke ordering

### DISCUSSION THREAD

**Discussion by rostero1** [statelyai/xstate#4104] [General]
In v5 I am trying to determine how best to set some data in context, so that it will be available to the invoked promise and throughout the parent.

If I set this data in `entry` and then try to set the context in `input`, `input` sees the context value as undefined, and of course the invoked promise then sees it as undefined.

I'm curious what the entry vs invoke order should be in v5.

**Comment by davidkpiano** [comment]
Can you share code to better show the problem you're facing?

**EDIT:** See comments for more details, this solution may not work for OP's use-case

In general, `input` provided to an invoked promise inside a machine can be done like this:

```ts
const machine = createMachine({
  context: ({ input }) => ({
    data: input, // transform/save input into machine context
  }),
  invoke: {
    src: fromPromise(async ({ input }) => {
      input; // { name: 'David' }
      // ...
    }),
    input: ({ context }) => context.data.user,
    // ...
  }
});

const actor = interpret(machine, {
  input: {
    user: { name: 'David' },
    foo: 'bar',
    // ...
  }
});

actor.start();
```

**Reply by Andarist** [reply]
The OP's problem is that `context` received by `input` is the one with values **pre** `entry` actions.
```ts
context: { count: 0 }
invoke: {
  src: 'someActor',
  input: ({ context }) => context.count // actual: 0, expected: 1
},
entry: assign({ count: 1 })
```

This is a change in v5, not necessarily a deliberate one though. I know about this subtle diff and was bringing this up on Discord [here](https://discord.com/channels/795785288994652170/800812262473728000/1009066035980292196)

**Reply by davidkpiano** [reply]
Ah okay. Invoked actors are _created_ (with `input`) before entry actions in v5, but they're not started until afterwards. Looking more closely, this may deviate from the SCXML order in [the algorithm](https://www.w3.org/TR/scxml/#AlgorithmforSCXMLInterpretation).

However, if we were to change this, then this other (expected?) use-case would break:

```ts
// ...
// Uh oh, 'actor' does not exist yet!
entry: sendTo('actor', { type: 'event' }),
invoke: {
  id: 'actor'
}
```

**Reply by Andarist** [reply]
yes, but I think that it's likely possible to fix this, quoting the mentioned thread:
> i think that it's best to run the actor factories~ after entry actions, so they would have access to the updated context
> but at the same time, sends should be buffered in a way that would make them delivered to those actors that are just going to be created

This is effectively what `deferredToBlockEnd` is about in v4:
https://github.com/statelyai/xstate/blob/a24b60b4a02f3296058f7fd850a4494d35763749/packages/core/src/actions.ts#L781-L782

**Reply by davidkpiano** [reply]
I'm okay with that; the only _slight_ issue is it will make it seem like `sendTo` actions are no longer in order, making action order more nuanced. So if we have an action order of this:

```ts
assign({ count: 0 }),
sendTo('actor', ({ context }) => ({ type: 'count0', val: context.count })),
assign({ count: 1 }),
sendTo('actor', ({ context }) => ({ type: 'count1', val: context.count })),
```

What is expected? As a user, regardless of order, I expect `count0` to send `0` and `count1` to send `1`.

I think we can still do this (i.e. execute actions in order) and defer the _sending_ of these events to the actors. What do you think?

**Reply by Andarist** [reply]
> I think we can still do this (i.e. execute actions in order) and defer the sending of these events to the actors. 

That's exactly what I'm proposing. Evaluation of "resolvers" shouldn't be affected, just the `lookupActor(mightBeAString).send(alreadyPreresolvedEvent)` should be "deferred" to the **entry block's end**
